File: Neetcode Questions/Linked List/add-two-numbers.cpp

// https://leetcode.com/problems/add-two-numbers/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:

    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        if(!l1 && !l2) return NULL;
        if(l1 && !l2) return l1;
        if(!l1 && l2) return l2;

        ListNode*head = NULL, *tail = NULL;
        int carry = 0;

        while(l1 && l2){
            int curr = l1->val + l2->val + carry;
            carry = curr/10;
            curr = curr%10;
            ListNode*newNode = new ListNode(curr);
            if(!head) head = tail = newNode;
            else{
                tail->next = newNode;
                tail = newNode;
            }
            l1 = l1->next;
            l2 = l2->next;
        }

        while(l1){
            int curr = l1->val + carry;
            carry = curr/10;
            curr = curr%10;
            ListNode*newNode = new ListNode(curr);
            tail->next = newNode;
            tail = newNode;
            l1 = l1->next;
        }

        while(l2){
            int curr = l2->val + carry;
            carry = curr/10;
            curr = curr%10;
            ListNode*newNode = new ListNode(curr);
            tail->next = newNode;
            tail = newNode;
            l2 = l2->next;
        }

        if(carry > 0){
            // carry = carry/10;
            ListNode*newNode = new ListNode(carry);
            tail->next = newNode;
            tail = newNode;
        }

        tail->next = NULL;

        // head = reverseLL(head);
        return head;
    }
};

File: Neetcode-150-js/Linked List/add-two-numbers.js

// https://leetcode.com/problems/add-two-numbers/description/

/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var addTwoNumbers = function(l1, l2) {
    if(!l1) return l2;
    if(!l2) return l1;
    if(!l1 && !l2) return null;

    let carry = 0;
    let head = null;
    let tail = null;

    while(l1 || l2 || carry){
        let val1 = l1 ? l1.val : 0;
        let val2 = l2 ? l2.val : 0;
        
        let currSum = val1 + val2 + carry;
        let num = currSum%10;
        carry = Math.floor(currSum/10);
        let newNode = new ListNode(num);
        if(!head){
            head = newNode;
            tail = newNode;
        } else{
            tail.next = newNode;
            tail = tail.next;
        }
        if(l1) l1 = l1.next;
        if(l2) l2 = l2.next;
    }

    return head;
};